### DISCUSSION TITLE
logrotate and copytruncate

### DISCUSSION THREAD

**Discussion by dodysw2** [martin-helmich/prometheus-nginxlog-exporter#235] [General]
Question: is it expected for logrotate to always use copytruncate option for exporter to be able to read new log file properly? I have a quite busy nginx that needs hourly rotation, on which copytruncate would be too disk i/o intensive as one hour worth of log file size is already a bit large. However the exporter almost always fail to realize the file has been rotated -- it's still holding the old file descriptor -- some times to the point of the file already deleted (exporter still kept holding on the fd). Using copytruncate fix this problem, but why is the exporter behave this way?

**Comment by martin-helmich** [comment]
Sorry for the late response.

This package uses the `github.com/hpcloud/tail` package for following log files, which also has some reported issues regarding log rotation (e.g. https://github.com/hpcloud/tail/issues/154, https://github.com/hpcloud/tail/issues/98 or https://github.com/hpcloud/tail/issues/94).

`github.com/hpcloud/tail` looks mostly abandoned, though. There seems to be a better-maintained fork (`github.com/nxadm/tail`), which apparently has addressed some of the open issues. I'll try bumping the package to its fork; possibly some of these issues will have already been fixed.
